TST: add broadcast_arrays() kwarg unit test for TypeError

tylerjereddy opened the pull request:
In short, local coverage tests (hopefully reported online soon from: #11567) indicated that we weren't unit testing for the proper handling of invalid kwargs for `np.broadcast_arrays()`.

In adding the unit test for this kwarg handling in this PR, I also noticed that the appropriate error **message** is not raised in Python 3 because `dict.keys()` is not a list anymore, so the test was also adjusted to verify the correct error message & the associated source code modified to ensure an indexable object in Python 3.

charris commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -57,6 +58,19 @@ def test_same():
+        broadcast_arrays(x, y, dtype='float64')
+
+    assert ('broadcast_arrays() got an unexpected keyword' in
+            str(excinfo.value))
```
Hmm, I think we are still using `assert_`. We should probably have a agreement about how to proceed with the new options pytest gives us.

tylerjereddy replied:
For reference, SciPy [had this discussion](https://github.com/scipy/scipy/issues/8583) a few months ago--not sure if we should similarly open an issue or ask on the mailing list? They decided in favor of allowing pytest `assert` usage in new PRs, but didn't mandate a mass migration.

Perhaps I'll just revert here for now?

eric-wieser replied:
If we were using pytest more broadly here, I think you could spell this `excinfo.match(r'broadcast_arrays\(\) got an unexpected keyword')`

charris commented on numpy/lib/tests/test_stride_tricks.py:
```diff
@@ -1,6 +1,7 @@
 from __future__ import division, absolute_import, print_function
 
 import numpy as np
+import pytest
```
We generally import system stuff first.

tylerjereddy commented:
I've restored usage of `assert_` to reduce the friction there.

As far as I can tell the NumPy `assert_raises` doesn't support catching the error string associated with an exception (or it isn't documented?), so if the use of `pytest.raises()` is not acceptable here I'll also have to remove the check for the error string / message added in testing changes here.

charris commented:
You probably want `assert_raises_regex`.

tylerjereddy commented:
`assert_raises_regex` calls `nose.tools.assert_raises_regex` in [the source](https://github.com/numpy/numpy/blob/v1.13.0/numpy/testing/utils.py#L1193-L1217)

charris commented:
@tylerjereddy `assert_raises-regex` no longer uses nose, it comes from unittest, which is where nose got it from. You are looking at old code.

tylerjereddy commented:
@charris Thanks for pointing out my mistake; I've now completely removed the `pytest` usage in the affected testing module and it seems to work just fine locally--hopefully the CI agrees.

tylerjereddy committed:
TST: add broadcast_arrays() kwarg unit test for TypeError

* broadcast_arrays() is now tested for the case when an invalid
keyword argument is used; the appropriate error string content
is also tested for

* the TypeError message produced in the above case has been restored
to the correct value in Python 3

eric-wieser committed:
MAINT: Use raw strings for regexes

eric-wieser reviewed (approved):
Made one small style tweak to reduce escaping - feel free to merge when tests pass

mattip commented:
Thanks @tylerjereddy
